### PR TITLE
Fix BN254 G1/G2 Point Encoding and Decoding

### PR DESCRIPTION
**Pull request by jayz22**
### What

Corrects the serialization format for BN254 G1 and G2 points to match [cap-74](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0074.md), which is modeled after EVM precompile specifications ([eip-196](https://eips.ethereum.org/EIPS/eip-196#encoding) and [197](https://eips.ethereum.org/EIPS/eip-197#encoding)):
  - All field elements now use big-endian encoding
  - Extension field elements (G2 points) are serialized with the imaginary component (c1) followed by the real component (c0)
  - The two flag bits are unset -- point-at-infinity is represented as `(0, 0)`, serialized as a zero-byte array; and a point with negative y component does not have the y-sign flag (bit 0) set. 
     
Additional testing:
- Added additional external hardcoded tests from https://www.evm.codes/precompiled
- More Serialization round-trip validation including invalid inputs and edge cases

**Acknowledgments**
Big thanks to @Oghma for testing the preview version, discovering this bug, and reporting it to us.

### CODE REVIEW

File: soroban-env-host/src/crypto/bls12_381.rs [hunk 1]
```diff
@@ -45,6 +45,8 @@ impl Host {
     // (the base field element), and will be charged accordingly.
     // Validation of the deserialized entity must be performed outside of this
     // function, to keep budget charging isolated.
+
+    // TODO: this function has to be moved into curve_utils because it is used by both
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-host/src/crypto/bls12_381.rs]: removed

File: soroban-env-host/src/crypto/bn254.rs [hunk 2]
```diff
@@ -36,30 +37,92 @@ impl Host {
...
-    pub(crate) fn bn254_deserialize_uncompressed_no_validate<
+    // Arkworks and ethereum differ in terms of how flags are handled.
+    //
+    // - In arkworks, the two free bits are reserved for flags: the
+    //   most-significant-bit is the Y-sign flag, and the 2nd
+    //   most-significant-bit is the infinity flag
+    // - In ethereum, the two free bits are always unset.
+    //
+    // Since encoding/decoding is in uncompressed mode (Y is included), there is
+    // no disambiguity. But we need to make sure to stick to the our
```
Comment (dmkozh) [inline soroban-env-host/src/crypto/bn254.rs]: typo: 'to our'

File: soroban-env-host/src/crypto/bn254.rs [hunk 3]
```diff
@@ -36,30 +37,92 @@ impl Host {
...
+        let flags = 0b1100_0000 & bytes[0];
+        if flags != 0b0000_0000 {
+            return Err(self.err(
+                ScErrorType::Crypto,
+                ScErrorCode::InvalidInput,
+                format!("bn254 {} deserialize: the two flag bits must be unset", tag).as_str(),
+                &[],
+            ));
+        }
+        let inf = bytes.iter().all(|b| *b == 0u8);
```
Comment (dmkozh) [inline soroban-env-host/src/crypto/bn254.rs]: Can we add some comment on infinity representation? For some reason I thought it was all 0's...

File: soroban-env-host/src/test/bn254.rs [hunk 4]
```diff
@@ -642,3 +645,472 @@ fn test_serialization_roundtrip() -> Result<(), HostError> {
...
+    Ok(())
+}
+
+#[test]
+fn test_bn254_g2_deserialize_rejects_y_sign_bit() -> Result<(), HostError> {
+    let host = observe_host!(Host::test_host());
+    host.enable_debug()?;
+    let mut rng = StdRng::from_seed([0u8; 32]);
+
+    // This is a negative g1 point serialized by us, it should not set the y-sign bit
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-host/src/test/bn254.rs]: fixed

File: soroban-env-host/src/test/bn254.rs [hunk 5]
```diff
@@ -76,6 +73,20 @@ fn compressed_g1(host: &Host, rng: &mut StdRng) -> Result<BytesObject, HostError
...
 }
 
+fn negative_g1(host: &Host, rng: &mut StdRng) -> Result<BytesObject, HostError> {
+    loop {
+        let g1 = G1Affine::rand(rng);
+        if g1.y().is_none() {
+            // infinity
+            continue;
+        }
+        if g1.y <= -g1.y {
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-host/src/test/bn254.rs]: fixed

File: soroban-env-host/src/crypto/bn254.rs [hunk 6]
```diff
@@ -36,30 +37,92 @@ impl Host {
...
+        return Ok(inf);
+    }
+
+    // The encoding/decoding of field elements (Fp and Fp2) differs from
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-host/src/crypto/bn254.rs]: yeah I meant arkworks and Ethereum handle them differently, we match Ethereum's. Cleaned up the wording.

File: soroban-env-host/observations/25/test__bn254__test_bn254_multi_pairing_check.json [hunk 7]
```diff
@@ -1,108 +1,216 @@
...
+  " 204 ret bytes_new_from_slice -> Ok(Bytes(obj#151))": "cpu:9165, mem:880, objs:-/76@833cb982",
+  " 205 call vec_new_from_slice(2)": "",
+  " 206 ret vec_new_from_slice -> Ok(Vec(obj#153))": "cpu:10254, mem:992, objs:-/77@33f4549b",
+  " 207 call bn254_multi_pairing_check(Vec(obj#147), Vec(obj#153))": "",
+  " 208 call vec_len(Vec(obj#147))": "cpu:10498",
+  " 209 ret vec_len -> Ok(U32(2))": "cpu:10620",
+  " 210 call vec_len(Vec(obj#153))": "cpu:11253, mem:1136",
+  " 211 ret vec_len -> Ok(U32(2))": "cpu:11375",
+  " 212 ret bn254_multi_pairing_check -> Ok(True)": "cpu:11408419, mem:100612",
+  " 213 end": "cpu:11408419, mem:100612, prngs:-/-, objs:-/77@33f4549b, vm:-/-, evt:-, store:-/-, foot:-, stk:-, auth:-/-"
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-host/observations/25/test__bn254__test_bn254_multi_pairing_check.json]: yes, the ~6M less CPU in the last case comes the fact that we now early-exit in the deserialization path, if the point is a zero, without doing subgroup check (~3M each)

File: soroban-env-common/env.json [hunk 8]
```diff
@@ -2367,7 +2367,7 @@
                         { "name": "point2", "type": "BytesObject" }
                     ],
                     "return": "BytesObject",
-                    "docs": "Adds two BN254 G1 points; encoding is 64-byte be_bytes(X)||be_bytes(Y); infinity is all-zeroes. No subgroup check.",
+                    "docs": "Adds two BN254 G1 points. G1 encoding: 64-byte uncompressed format: be_bytes(X)||be_bytes(Y), where X and Y are 32-byte big-endian Fp field elements. The two flag bits (bit 0 and 1) of the first byte must be unset -- infinity is represented as 64 zero bytes. Points must be on curve.",
```
[the remark this answers is not in the document]
Comment (jayz22) [reply soroban-env-common/env.json]: It was correct. Subgroup check is not needed because it's always true. I've added it back for clarity.
